### PR TITLE
Parse URDF joint hard limits into the HardwareInfo structure

### PR DESCRIPTION
**Pull request by saikishor**
This PR allows to parse the joint_limits into the `HardwareInfo` structure, so that this can be later reused by the ResourceManager to enforce the limits on the `JointCommandInterfaces`.

The joint limits that are available in the `HardwareInfo` structure are basically parsed from the URDF joint limits, and these joint limits range can be reduced with the `<min>` and `<max>` tag of each command interface, the idea is to be able to always be able to conservative.

There will be a following PR on parsing the `SoftLimits` as well

### CODE REVIEW

**Review by christophfroehlich — CHANGES_REQUESTED** [review]
The idea is clear and the implementation looks fine. However, a couple of lines [miss coverage](https://app.codecov.io/gh/ros-controls/ros2_control/pull/1472/blob/hardware_interface/src/component_parser.cpp), especially in the `update_interface_limits` lambda

**Review by christophfroehlich — COMMENTED** [review]
10 lines are still not covered, and total coverage will decrease. should I try to add tests to run into this failure cases?

**Comment by saikishor** [comment]
> 10 lines are still not covered, and total coverage will decrease. should I try to add tests to run into this failure cases?

Sure. if you can, it would be really helpful

**Review by olivier-stasse — APPROVED** [review]

File: hardware_interface/src/component_parser.cpp
Comment (christophfroehlich) [reply hardware_interface/src/component_parser.cpp:725]: what do you mean by default initialization? `limits` will be filled from the URDF tag, which is mandatory for REVOLUTE and PRISMATIC joints -> max_velocity is already populated

Comment (saikishor) [reply hardware_interface/src/component_parser.cpp:790]: Yes, moreover, there might be a joint without velocity limits, so it doesn't make sense

File: hardware_interface/src/component_parser.cpp
Comment (saikishor) [reply hardware_interface/src/component_parser.cpp:803]: Well same as above

File: hardware_interface/src/component_parser.cpp
Comment (saikishor) [reply hardware_interface/src/component_parser.cpp:846]: Jerk limits are very familiar as the support in ROS is little, so I think it is better to leave it up to the user

**Comment by saikishor** [comment]
@christophfroehlich Thanks for helping out and adding very nice tests

**Review by christophfroehlich — DISMISSED** [review]
Just let us add some notes about this once #1477 is merged.

**Comment by saikishor** [comment]
> Just let us add some notes about this once #1477 is merged.

Sure. I will do that :)

File: hardware_interface/src/component_parser.cpp
Comment (bmagyar) [inline hardware_interface/src/component_parser.cpp:704]: I noticed this pattern of `true &&` in a couple of places. Why?

File: hardware_interface/src/component_parser.cpp
Comment (bmagyar) [inline hardware_interface/src/component_parser.cpp:657]: These lambdas are way too large for my taste. Could you please convert them to member functions? At least some of them...

Comment (saikishor) [reply hardware_interface/src/component_parser.cpp:657]: Sure, I can do that

File: hardware_interface/src/component_parser.cpp
Comment (saikishor) [reply hardware_interface/src/component_parser.cpp:704]: You are right, it can be simply `itr.enable_limits`. I will make the modifications soon and let you know

File: hardware_interface/src/component_parser.cpp
Comment (saikishor) [reply hardware_interface/src/component_parser.cpp]: Done

File: hardware_interface/src/component_parser.cpp
Comment (saikishor) [reply hardware_interface/src/component_parser.cpp]: You are right. This is fixed now. :+1:

File: hardware_interface/src/component_parser.cpp
Comment (bmagyar) [reply hardware_interface/src/component_parser.cpp]: thank you!

**Review by bmagyar — APPROVED** [review]

**Pull request merged**
